### PR TITLE
Add configuration for application environment

### PR DESCRIPTION
**Pull request by marcotc**
This new configuration option ensures that traces and metrics share the same `env` tag, allowing proper correlation between them during back-end processing.

This configuration is done at application configuration level because it affects more than just the tracer: it applies to health metrics, runtime metrics and traced spans.

### CODE REVIEW

File: docs/GettingStarted.md [hunk 1]
```diff
@@ -1455,19 +1455,31 @@ Datadog::Tracer.log.info { "this is typically called by tracing code" }
 
 ### Environment and tags
 
-By default, the trace agent (not this library, but the program running in the background collecting data from various clients) uses the tags set in the agent config file, see our [environments tutorial](https://app.datadoghq.com/apm/docs/tutorials/environments) for details.
+By default, the trace agent (not this library, but the program running in the background collecting data from various clients) uses the tags set in the agent config file, see our documentation around [Assigning Tags](https://docs.datadoghq.com/tagging/assigning_tags/?tab=agentv6#add-tags) and [APM environment configuration](https://docs.datadoghq.com/tracing/send_traces/#configure-your-environment) for details.
```
Comment (marcotc) [inline docs/GettingStarted.md]: Previous link was broken: that page has been completely redone.

File: lib/ddtrace/metrics.rb [hunk 2]
```diff
@@ -49,13 +56,14 @@ def default_statsd_client
...
       self.enabled = options[:enabled] if options.key?(:enabled)
+      @env = options[:env] if options.key?(:env)
     end
 
     def send_stats?
       enabled? && !statsd.nil?
     end
 
-    def count(stat, value = nil, options = nil, &block)
+    def count(stat, value = nil, options = {}, &block)
```
Comment (marcotc) [inline lib/ddtrace/metrics.rb]: The default value for `options` has been changed to `{}` to allow for a much easier merging of options in `metric_options`.

File: lib/ddtrace/metrics.rb [hunk 3]
```diff
@@ -122,6 +130,26 @@ def initialize(*args)
...
+      # Return dupes, so that the constant isn't modified,
+      # and defaults are unfrozen for mutation in Statsd.
+      super.tap do |options|
+        options[:tags] = options[:tags].dup
+
+        # Add tags dynamically because they might change during
+        # runtime (i.e. when calling #configure).
+        options[:tags].concat(instance_tags)
+      end
+    end
```
Comment (marcotc) [inline lib/ddtrace/metrics.rb]: This logic is very similar to what we do in `Datadog::Runtime::Metrics`: https://github.com/DataDog/dd-trace-rb/blob/bcfeb95d21e3c0a2ded57388d6af0e23597215be/lib/ddtrace/runtime/metrics.rb#L69-L78

File: lib/ddtrace/metrics.rb [hunk 4]
```diff
@@ -122,6 +130,26 @@ def initialize(*args)
...
 
+    # Return instance specific tags.
+    def instance_tags
+      tags = []
+      tags << "#{Ext::Metrics::TAG_ENV}:#{env}" if env
+      tags
+    end
+
+    # Add instance tags to default metrics
+    def default_metric_options
```
Comment (delner) [inline lib/ddtrace/metrics.rb]: Why rewrite this method? Could you simply add `options[:tags].concat(instance_tags)` to the existing method?

File: lib/ddtrace/metrics.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/metrics.rb]: That's a fair point, I'll try to implement this in `metric_options` once again.

File: lib/ddtrace/metrics.rb (hunk 4 above)
Comment (marcotc) [reply lib/ddtrace/metrics.rb]: I was able to simplify this logic, thank you for the feedback.

**Comment by marcotc** [comment]
@delner comments addressed! Feel free to take a look again.

**Comment by delner** [comment]
@marcotc I think this is obsolete via #985 given we now have `env` in our configuration settings. Should we close this?

**Comment by marcotc** [comment]
@delner I believe you are correct here, I'll close it!

**Pull request closed without merging**